operator: fix crash that occurs when cep CRD is disabled and CES enabled with kvstore

doniacld opened the pull request:
## Description
Make the operator crash early if CiliumEndpointSlice is enabled and CiliumEndpoint CRDs is disabled since it does not make sense to have CES without cep running. 

Add a check at helm installation level.

Fixes #24396 

## Test

### Test at runtime

Adding the following configuration in kind-values.yaml and deploy a kind cluster
```yaml
enableCiliumEndpointSlice: true
disableEndpointCRD: true
identityAllocationMode: "kvstore"
```
NB: Tested with and without `identityAllocationMode: "kvstore"`

Check the logs from the operator
```
$ ks get pod                                                                                    
NAME                                         READY   STATUS             RESTARTS     AGE
cilium-24plg                                 0/1     Init:3/6           0            3s
cilium-fblnk                                 0/1     Init:3/6           0            3s
cilium-operator-65fb77cf85-jv8xh             0/1     CrashLoopBackOff   1 (2s ago)   3s
```

```
$ ks logs cilium-operator-65fb77cf85-jv8xh -f                                                                  
level=debug msg="Skipped reading configuration file" error="Config File \"cilium-operators\" Not Found in \"[/]\"" subsys=config
level=fatal msg="Running Cilium with enable-cilium-endpoint-slice=true requires disable-endpoint-crd set to false to enable CRDs." subsys=config
```

### Test during deployment

The error is properly raised.
```
$ helm install  --namespace kube-system    -f ./contrib/testing/kind-values.yaml cilium ./install/kubernetes/cilium   
Error: INSTALLATION FAILED: execution error at (cilium/templates/validate.yaml:76:7): Cilium operator requires .Values.enableCiliumEndpointSlice=true and .Values.disableEndpointCRD=false
```

```release-note
Change enableEndpointCRD helm option type from string to boolean
Fix operator panic that occurs when Endpoint CRD is disabled and CiliumEndpointSlice is enabled
```

doniacld committed:
operator: fix crash when cep CRD is disabled, CES enabled with
kvstore

Make the operator crash early in config.go if CiliumEndpointSlice is enabled
and CiliumEndpointCRD is disabled. Note that CiliumEndpointSlice feature needs
CiliumEndpoint CRDs to run.  In operator/cmd/root.go, remove the condition on
CiliumEndpointCRD since at this point of the lifecycle, the operator should
have crash if we end up in this case. This fix is only relevant for kvstore
mode, because in identity-allocation-mode CRD, we forcefully enable the
CiliumEndpoint CRD.

Fixes #24396

Signed-off-by: Donia Chaiehloudj <donia.cld@isovalent.com>

mhofstetter reviewed (commented):
hey @doniacld 

thanks for the PR - looks good so far! only some proposals regarding the failing checks and missing labels from my side.

**Checks**

looks like the check `BPF checks / checkpatch` is failing due to the long commit message subject.

`Error: ERROR:CUSTOM: Please avoid long commit subjects (max: 75, found: 85)`

Not sure what's going on with the other failing `CIlium Runtime` checks that have been introduced lately. Might be worth to rebase to `main` to fetch the latest version while renaming the commit message.

**Labels**

you should also add a relevant release note label to this PR (https://docs.cilium.io/en/latest/contributing/development/contributing_guide/#submitting-a-pull-request - 11.) - IMO `release-note/misc` should be appropriate in this case (which doesn't require an explicit release note).

it's also helpful to add some labels to bring the PR into context - e.g. `area/operator`, `kind/bug`

mhofstetter reviewed (commented):
@doniacld thanks for adding the labels and fixing the commit subject.

Check in the Go-Code when starting the operator looks good to me.

But it might be worth to add this check to the Helm Chart too - to prevent an installation with these incompatible values in an earlier stage. This would also cover an installation via Cilium CLI (with helm mode).

The two config values are configured in the configmap ([disableEndpointCRD](https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-configmap.yaml#L131-L133) & [enableCiliumEndpointSlice](https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-configmap.yaml#L985-L987)). 

I'd recommend to place a check based on the two helm values `.Values.enableCiliumEndpointSlice` & `.Values.disableEndpointCRD` in the helm validation file (https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/validate.yaml)

doniacld committed:
helm: Check EndpointCRD is enabled when CiliumEndpointSlice is enabled too

This commit adds a supplementary check during the deployment to prevent
installing Cilium with incompatibles flags enableCiliumEndpointSlice=true
disableEndpointCRD=true.

Signed-off-by: Donia Chaiehloudj <donia.cld@isovalent.com>

kaworu commented on install/kubernetes/cilium/templates/validate.yaml:
```diff
@@ -69,3 +69,10 @@
+{{/* validate Cilium operator */}}
+{{- if .Values.enableCiliumEndpointSlice }}
+  {{- if .Values.disableEndpointCRD }}
+    {{ fail "if enableCiliumEndpointSlice is true, that then disableEndpointCRD needs to be false" }}
```
nitpicking

```suggestion
    {{ fail "Cilium Endpoint Slice (.Values.enableCiliumEndpointSlice=true) requires .Values.disableEndpointCRD=false" }}
```

doniacld replied:
Updated.

kaworu reviewed (approved):
Thanks @doniacld patch LGTM!

mhofstetter reviewed (approved):
@doniacld thanks for adding the helm validation!

looks like the docker build failed due to some infra issues (and therefore all dependent checks).

i'd recommend to rebase to `main` to re-trigger all basic checks. In addition you can trigger the e2e-tests by commenting `/test` (in a single comment)

@tommyp1ckles your review became mandatory (with sig-k8s). please chime in :)

doniacld committed:
helm: Change disableEndpointCRD type from string to boolean

Signed-off-by: Donia Chaiehloudj <donia.cld@isovalent.com>

qmonnet reviewed (approved):
Looks good, thank you!

mhofstetter commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -131,9 +131,9 @@ data:
 {{- end }}
 
-{{- if hasKey .Values "disableEndpointCRD" }}
+{{- if .Values.disableEndpointCRD }}
```
to enforce an explicit type check - we have to change this to `{{- if eq .Values.disableEndpointCRD true }}`

this will result in an error if trying to pass an incompatible type.

```
error calling eq: incompatible types for comparison
```

doniacld replied:
Updated and tested.

### Configuration with the wrong type
```
enableCiliumEndpointSlice: false
disableEndpointCRD: "true"
```
Error is as expected
```
Error: template: cilium/templates/cilium-configmap.yaml:134:7: executing "cilium/templates/cilium-configmap.yaml" at <eq .Values.disableEndpointCRD true>: error calling eq: incompatible types for comparison
```

### Configuration with the wrong type BUT enableCiliumEndpointSlice: true
```
enableCiliumEndpointSlice: true
disableEndpointCRD: "true"
```
The error type is not thrown since I guess validate.yaml is the first test to be executed.
```
Error: execution error at (cilium/templates/validate.yaml:76:7): if Cilium Endpoint Slice is enabled (.Values.enableCiliumEndpointSlice=true), it requires .Values.disableEndpointCRD=false
```

mhofstetter replied:
@doniacld great

we're missing one case:

```
enableCiliumEndpointSlice: true
disableEndpointCRD: "false"
```

the validation error gets thrown even though the endpointCRD isn't actually disabled 

```
❯ helm template ./install/kubernetes/cilium -f ./contrib/testing/kind-values.yaml
Error: execution error at (cilium/templates/validate.yaml:76:7): if Cilium Endpoint Slice is enabled (.Values.enableCiliumEndpointSlice=true), it requires .Values.disableEndpointCRD=false
```

-> we should change the validation to enforce type checks too

```
{{- if and (eq .Values.enableCiliumEndpointSlice true ) (eq .Values.disableEndpointCRD true ) }}
  {{ fail "if Cilium Endpoint Slice is enabled (.Values.enableCiliumEndpointSlice=true), it requires .Values.disableEndpointCRD=false" }}
{{- end }}
```

doniacld replied:
Little update:

### Config enableCiliumEndpointSlice:false & disableEndpointCRD: "true"
```yaml
enableCiliumEndpointSlice: false
disableEndpointCRD: "true"
```

Error
```yaml
Error: template: cilium/templates/cilium-configmap.yaml:134:7: executing "cilium/templates/cilium-configmap.yaml" at <eq .Values.disableEndpointCRD true>: error calling eq: incompatible types for comparison
```

### Config enableCiliumEndpointSlice:true & disableEndpointCRD: "true"
```yaml
enableCiliumEndpointSlice: true
disableEndpointCRD: "true"
```
Error
```
Error: template: cilium/templates/validate.yaml:75:9: executing "cilium/templates/validate.yaml" at <eq .Values.disableEndpointCRD true>: error calling eq: incompatible types for comparison
```

### Helm command enableCiliumEndpointSlice=true & disableEndpointCRD="true"
```
 --set enableCiliumEndpointSlice=true --set-string disableEndpointCRD="true"
```
Error
```
Error: template: cilium/templates/validate.yaml:75:9: executing "cilium/templates/validate.yaml" at <eq .Values.disableEndpointCRD true>: error calling eq: incompatible types for comparison
```

Now we should be good!

mhofstetter replied:
thanks @doniacld
